### PR TITLE
Traditional Chinese translation of "Why Do We Write super(props)?"

### PR DESCRIPTION
**Pull request by qas612820704**
Hi Dan, your articles are very valuable, I learned a lot. Thanks for sharing. 

This pull request contains a translation of the article "Why Do We Write super(props)?" to Traditional Chinese.

I have followed [中文文案排版指北](https://github.com/sparanoid/chinese-copywriting-guidelines) this Chinese copywriting guidelines.

If there anywhere needs to be fixed, let me know. 

Again, thanks!!

### CODE REVIEW

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+沒錯，[類別欄位提案](https://github.com/tc39/proposal-class-fields) 讓我們能省略宣告：
+
+```jsx
+class Checkbox extends React.Component {
+  state = { isOn: true };
+  // ...
+}
+```
+
+在 2015 年， React 0.13 新增對一般類別的支援，[計劃](https://reactjs.org/blog/2015/01/27/react-v0.13.0-beta-1.html#es7-property-initializers) 藉由定義 `constructor` 和呼叫 `super(props)` 的語法當作臨時的解決方法，直到類別欄位提供合適的替代方案。
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
類似這樣的語法是在 2015 年React 0.13 新增對一般類別的支援時就有的[計劃](https://reactjs.org/blog/2015/01/27/react-v0.13.0-beta-1.html#es7-property-initializers) 。
。定義 `constructor` 和呼叫 `super(props)` 的語法一直都被當作是暫時的解決方法，直到類別欄位提供合適的替代方案。
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    // 🔴 還不能用 `this`
+    super(props);
+    // ✅ 現在沒問題了
+    this.state = { isOn: true };
+  }
+  // ...
+}
+```
+
+這有一個很好的理由說明，為什麽 JavaScript 會強制父建構子在你碰 `this` 前被執行。想想類別的階層：
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
JavaScript 會強制父建構子在你碰 `this` 前被執行是有原因的。想想類別的階層：
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    this.greetColleagues(); // 🔴 這不被允許，看下面說明
+    super(name);
+  }
+  greetColleagues() {
+    alert('Good morning folks!');
+  }
+}
+```
+
+想像一下，如果 `super` 被呼叫前使用 `this` *是*允許的。幾個月後，我們可能想讓人（person）的名子（name）包含在 `greetColleagues` 的訊息中：
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
想像一下如果在 `super` 被呼叫前使用 `this` *是*被允許的會是怎樣的情況。幾個月後，我們可能想讓某人的名
字（name）包含在 `greetColleagues` 的訊息中：
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+```jsx
+  greetColleagues() {
+    alert('Good morning folks!');
+    alert('My name is ' + this.name + ', nice to meet you!');
+  }
+```
+
+但是我們是否忘了，呼叫 `this.greetColleagues()` 是在 `super()` 有機會設置 `this.name` 之前。所以 `this.name` 根本不曾被定義過！就跟你看到的一樣，可能很難想到會有像這樣的程式碼。
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
但是我們忘了 `this.greetColleagues()` 是在 `super()` 有機會設置 `this.name` 之前就被呼叫了。所以 `this.name` 根本不曾被定義過！如你所見，這樣的程式碼是很難理解的。
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  constructor(props) {
+    this.props = props;
+    // ...
+  }
+}
+```
+
+這離真實情況相去不遠 — 的確，它是 [這麼做](https://github.com/facebook/react/blob/1d25aa5787d4e19704c049c3cfa985d3b5190e0d/packages/react/src/ReactBaseClasses.js#L22) 的。
+
+但不知為何，即使你呼叫的是沒有帶入 `props` 參數的 `super()`，你仍然可以在 `render` 中或其他函式讀取 `this.props`。（如果你不相信我，自己試！）
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
但不知為何，即使你呼叫的是沒有帶入 `props` 參數的 `super()`，你仍然可以在 `render` 中或其他函式讀取 `this.props`。（如果你不相信我，自己試試看！）
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+那*它*是如何運作的？事實上 **在 React 呼叫*你的*建構子後，它也會馬上配置 `props` 到實例中：**
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
那*這*是如何運作的？事實上 **在 React 呼叫*你的*建構子後，它也會馬上配置 `props` 到實例中：**
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+```jsx
+  // React 內部
+  const instance = new YourComponent(props);
+  instance.props = props;
+```
+
+這就是為什麼就算你忘記帶入 `props` 到 `super()` 中，React 仍會在之後配置它。
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
這就是為什麼就算你忘記帶入 `props` 到 `super()` 中，React 仍會在之後配置它。這是有原因的。
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+當 React 增加對類別的支援，它並不是只想增加對 ES6 類別的支援。它的目標是盡可能廣泛地支援抽象的類別。目前還 [不明確](https://reactjs.org/blog/2015/01/27/react-v0.13.0-beta-1.html#other-languages) 如 ClojureScript，CoffeeScript，ES6，Fable，Scala.js，TypeScript 或其他的解決方案如何相對地成功定義元件，所以 React 故意不在乎是否必須呼叫 `super()` — 即是是 ES6 的類別也是如此。
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
當 React 增加對類別的支援，它並不是只想增加對 ES6 類別的支援。它的目標是盡可能廣泛地支援抽象的類別。當時
我們還 [不清楚](https://reactjs.org/blog/2015/01/27/react-v0.13.0-beta-1.html#other-languages) 如  ClojureScript，CoffeeScript，ES6，Fable，Scala.js，TypeScript 或其他的解決方案如何相對地成功定義元件，所以 當時 React 是刻意設計成對是否必須呼叫 `super()` 這點睜一隻眼閉一隻眼的 —— 即使是 ES6 的類別也是如此。
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+所以意思是說你可以只寫 `super()` 而不是 `super(props)`？
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
所以意思是說你可以只寫 `super()` 而不用寫 `super(props)` 嗎？
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+**最好不要，這仍然會造成誤解。**沒錯，React 會在你的建構子執行之後配置 `this.props`。但是 `this.props` *從*呼叫 `super` 到建構子的結尾仍然會是未定義（undefined）。
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
**最好不要，因為這仍然會造成誤解。**沒錯，React 會在你的建構子執行之後自行配置 `this.props`。但是 `this.props` *從*呼叫 `super` 到建構子的結尾仍然會是未定義（undefined）。
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    // ...
+  }
+}
+
+// 你的程式碼內部
+class Button extends React.Component {
+  constructor(props) {
+    super(); // 😬 我們忘了帶入 props
+    console.log(props);      // ✅ {}
+    console.log(this.props); // 😬 undefined
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
    console.log(this.props); // 😬 未定義（undefined）
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+  }
+  // ...
+}
+```
+
+如果某些*從*建構子中被呼叫的函式發生這種狀況，除錯將會更加艱難。**這就為什麽即使它不是絕對的必要，但我建議總是帶入 `super(props)`：**
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
如果某些*從*建構子中被呼叫的函式發生下面這種狀況，除錯將會更加艱難。**這就是為什麽我建議開發者總是要帶入 `super(props)`，即使這並非絕對必要：**
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+最後還有一點是長期以來 React 的使用者可能會感到好奇的。
+
+你或許有注意到，當你在類別中使用 Context API（不論是舊有 `contextTypes` 或是從 React 16.6 新增的新式 `contextType` API），`context` 會作為第二個參數傳給建構子。
+
+所以為什麼我們不取而代之寫成 `super(props, context)` 呢？其實我們可以，只是 context 使用的頻率比較低，會踩到的雷並不是那麼多。
+
+**當有了類別欄位提案，整個這種的雷大部份都會消失。**在沒有標明建構子的狀況下，所有的參數都會自動地被帶入。這樣就允許了像 `state = {}` 這樣的表達式，在有需要的狀況下，還是能包含參考（reference） `this.props` 跟 `this.context` 的能力。
+
+而當有了 Hooks 後，我們甚至不需要 `super` 或是 `this`。但這就是改天的主題了。
```
Comment (chloewlin) [inline src/pages/why-do-we-write-super-props/index.zh-hant.md]:
```suggestion
而當有了 Hooks 後，我們甚至不需要 `super` 或是 `this`。但這就改天再說了。
```

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
+---
+title: 為什麼我們要寫 super(props)?
```
[the remark this answers is not in the document]
Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Use fullwidth question mark

File: src/pages/why-do-we-write-super-props/index.zh-hant.md [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+date: '2018-11-30'
+spoiler: 在結尾有一個轉折。
+---
+
+
+我聽說 [Hooks](https://reactjs.org/docs/hooks-intro.html) 正夯， 但我想要從揭開 *class* 有趣的面紗來開始這個部落格。這樣如何！
+
+**這些解釋對於有效地使用 React 並*不*重要，但對於喜歡深入了解其運作的你，將會發現當中有趣的事情。**
```
[the remark this answers is not in the document]
Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Gotchas = pitfalls

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 1 above)
Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: This entire passage was in written past tense to explain the intentions back in 2015

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 8 above)
Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: This was in past tense

**Comment by chloewlin** [comment]
Hi @qas612820704 ! I proofread your translation and added some feedback. Feel free to tag me if you have any questions.

**Comment by chloewlin** [comment]
No Problem, @gaearon . You can count on me!

**Comment by qas612820704** [comment]
Hi @chloewlin, thanks you very much for helping me with this.
I will improve asap.

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (qas612820704) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Hi @chloewlin, Is "space" before the fullwidth question mark necessary?

[the remark this answers is not in the document]
Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Good catch. I think the whitespace is not necessary here (but you might need it if super(props) is surrounded by back ticks).

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 3 above)
Comment (qas612820704) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: @chloewlin, is there no need to remark "person" and "names" here?

They are the variable in the example below.

And there has a mistype newline after "名".

Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: I think the fact that they refer to `this.person` and `this.name` is obvious to anyone who reads the code snippet. It's also not in the original text. But maybe some reader will find it useful. It's a stylistic choice. :)

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 8 above)
Comment (qas612820704) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: @chloewlin, is there a space before "當時" ?

Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: No. See [here](https://github.com/sparanoid/chinese-copywriting-guidelines#%E5%85%A8%E5%BD%A2%E6%A8%99%E9%BB%9E%E8%88%87%E5%85%B6%E4%BB%96%E5%AD%97%E7%AC%A6%E4%B9%8B%E9%96%93%E4%B8%8D%E5%8A%A0%E7%A9%BA%E6%A0%BC).

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (qas612820704) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Okay, I got it! Thanks you point out :)

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 3 above)
Comment (qas612820704) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Yeah, you make sense. It's really good to read without this.

File: src/pages/why-do-we-write-super-props/index.zh-hant.md (hunk 8 above)
Comment (qas612820704) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: I mean the space between `所以` and `當時`.

Comment (chloewlin) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Ah, no. I removed it.

Comment (qas612820704) [reply src/pages/why-do-we-write-super-props/index.zh-hant.md]: Thanks you :)

**Comment by qas612820704** [comment]
I improve the translation from review, and squashing the commits.

@chloewlin , thanks for your correcting. I don't know how to express my gratitude to you.

If there are somewhere still need to be improved. Let me know, thanks!

**Comment by gaearon** [comment]
Let's get this in — @chloewlin feel free to make edits on top if you'd like!

**Pull request merged**